Load cells at xlsx worksheet ingest; handle skipping and/or blank row…

jennybc opened the pull request:
…s; fixes #224, fixes #163 

Partially closes #222 (skipping), which will remain open to cover re: xls. Other linked issues are mentioned in NEWS and commit messages.

hadley commented on src/CellType.h:
```diff
@@ -142,6 +142,7 @@ inline Rcpp::RObject makeCol(CellType type, int n) {
   switch(type) {
   case CELL_BLANK:
     return R_NilValue;
+    break;
```
Doh

jimhester replied:
These breaks aren't actually needed, if the function is going to return first you don't need to break out of the switch.

hadley replied:
Either way it seems like we should consistently use or not use in a given switch.

jennybc replied:
I wondered why there had been no suffering due this missing `break`. I will leave it because at least it's now all consistent.

hadley commented on src/XlsxWorkSheet.cpp:
```diff
@@ -47,6 +47,11 @@ List read_xlsx_(std::string path, int sheet, RObject col_names,
 
+  if (ws.nrow() == 0 && ws.ncol() == 0) {
+    Rcpp::List empty(0);
+    return empty;
```
You could just do `return Rcpp::List(0)`  

(And I assume this gets tibble-ised elsewhere?)

jennybc replied:
Aha! And YES re: tibble-isation.

hadley commented on src/XlsxWorkSheet.cpp:
```diff
@@ -58,7 +63,7 @@ List read_xlsx_(std::string path, int sheet, RObject col_names,
     sheetHasColumnNames = as<bool>(col_names);
     if (sheetHasColumnNames) {
-      colNames = ws.colNames(nskip);
+      colNames = ws.colNames(std::max(ws.min_row(), nskip));
```
What's the logic here?

hadley replied:
Since you use it in a few places, maybe assign to an informatively named variable?

jennybc replied:
I will see if I can remove or generalize `min_row_` now.

I think it might be wise to record info about where data starts (after advancing past `nskip`) and where the next row starts (assuming we'll need to advance past column names). Because yes I currently do this in at least two different places with the same logic.

hadley replied:
Maybe call it `first_row_`? I think that's a bit more descriptive.

And `first_data_row_`?

jennybc replied:
At the time of worksheet construction, right after dimensions are cached, I now mark the first cell with declared row >= `nskip` and the first cell with declared row >= to that of the former. Then all subsequent functions (column names, column types, column reading) just consult that.

hadley commented on src/XlsxWorkSheet.cpp:
```diff
@@ -72,11 +77,16 @@ List read_xlsx_(std::string path, int sheet, RObject col_names,
 
+  // if we know there's no data reading to be done, we should return a
+  // 0 row tibble HERE with the correct colNames
+  // TO THINK: case where colTypes were explicitly specified but no data to read
```
I think you could return a data frame with a 0-length vector for each col, but it's not necessary and few people will notice the omission

jennybc replied:
Done.

hadley commented on src/XlsxWorkSheet.h:
```diff
@@ -16,11 +16,14 @@
 class XlsxWorkSheet {
   XlsxWorkBook wb_;
+  // JB QUESTION: seems like a worksheet should know its own name (and/or
+  // position). This could be used in messaging, such as, "SHEET has no data"
```
Yes, I think that's a good idea

jennybc replied:
Still need to do this.

jennybc replied:
Done.

hadley commented on src/XlsxWorkSheet.h:
```diff
@@ -52,6 +56,13 @@ class XlsxWorkSheet {
     return nrow_;
   }
 
+  int min_row() {
```
`const`?

jennybc replied:
gone now anyway

hadley commented on src/XlsxWorkSheet.h:
```diff
@@ -52,6 +56,13 @@ class XlsxWorkSheet {
+    return min_row_;
+  }
+
+  // JB: this seems to be leftover from previous development?
```
I think you can/should kill it now that we have `loadCells()`

jennybc replied:
deleted this

hadley commented on src/XlsxWorkSheet.h:
```diff
@@ -66,31 +77,53 @@ class XlsxWorkSheet {
-        if (xcell.col() >= ncol_)
-          continue;
+    // go to row nskip or, if impossible, the next row that exists
+    std::vector<XlsxCell>::const_iterator it = getNextRow(nskip);
```
I think this should get moved out too. Maybe `loadCells()` should record which is the first non-blank cell that it sees?

hadley replied:
And maybe also record the first cell after all the skipping?

jennybc replied:
Agreed and acknowledged in above comment.

jennybc replied:
Deleted, in favour of marking the cells to start reading from at the time of ingest.

hadley commented on src/XlsxWorkSheet.h:
```diff
@@ -66,31 +77,53 @@ class XlsxWorkSheet {
+    int base = it->row();
+    // advance past column names and any embedded blank rows
+    if (has_col_names) {
+      it = getNextRow(base + 1);
```
It seems a bit inefficient to start again from scratch to find the next row. Maybe `getNextRow()` should take a ` std::vector<XlsxCell>::const_iterator` and a number of rows to advance?

jennybc replied:
But you don't know how many rows to advance. However, I think this will all go away anyway, once I record these "milestones" exactly once at the start of worksheet ingest.

jennybc replied:
I do this row identification exactly once now and then just work with the iterators.

hadley commented on src/XlsxWorkSheet.h:
```diff
@@ -99,54 +132,79 @@ class XlsxWorkSheet {
-      if (xcell.col() >= ncol_)
-        continue;
+    // go to row nskip or, if impossible, the next row that has data
+    std::vector<XlsxCell>::const_iterator it = getNextRow(nskip);
```
So does `nskip` include or exclude completely blank rows?

hadley commented on src/XlsxWorkSheet.h:
```diff
@@ -201,61 +261,90 @@ class XlsxWorkSheet {
+      // JB: it feels like I should reveal sheet ?name? here
+      // currently worksheet doesn't know it's own name, though
+      // Or perhaps I should say nothing at all?
+      Rcpp::warning("No <row> in sheet xml. No data read.");
```
I think generally loading an empty file should not generate a warning/message.

jennybc replied:
warning deleted

jennybc commented:
@hadley 

I think I've addressed all of your comments and have gotten the NEWs/issues sorted out. Once you are satisfied, I think this can be merged. There are still a few "to do's" but they are best addressed as part of the next phase, where I'll work on column names and types for xlsx.

hadley commented on src/XlsxWorkSheet.h:
```diff
@@ -20,28 +20,41 @@ class XlsxWorkSheet {
 
-  XlsxWorkSheet(XlsxWorkBook wb, int sheet_i): wb_(wb) {
+  XlsxWorkSheet(XlsxWorkBook wb, int sheet_i, int nskip):
+  wb_(wb)
```
I think our policy is to indent default constructors. @jimhester what do you do?

jimhester replied:
I [indent them](https://github.com/rstats-db/odbc/blob/71da07fdf51b30043edaf94c094f2da497fae882/src/odbc_result.h#L31-L36).

It might be worth standardizing on [clang-format](https://clang.llvm.org/docs/ClangFormat.html) and setting up a [`.clang-format`](https://github.com/lexicalunit/nanodbc/blob/master/.clang-format) file, which would give us standardized formatting across projects.

jennybc replied:
This seems not to be what RStudio wants to do. Does it do so for either of you and I've got some setting wrong?

I'm worried about doing it manually, because I use "reindent lines" so often, I fear my indenting will be worse overall, i.e. not internally consistent.

@jimhester Would RStudio implement a ` .clang-format` file?

hadley commented on src/XlsxWorkSheet.h:
```diff
@@ -52,101 +65,153 @@ class XlsxWorkSheet {
-  void printCells() {
-    for (rapidxml::xml_node<>* row = sheetData_->first_node("row");
-         row; row = row->next_sibling("row")) {
+  std::string sheetName() {
```
`const` ?

jennybc replied:
Is there any reason to not do same for `ncol()` and `nrow()`, just above this?

I added for all 3, so speak up if it's a mistake to add for the other two.

hadley commented on tests/testthat/test-empty.R:
```diff
@@ -0,0 +1,29 @@
+  ## currently we get a 0-row tibble with 2 unnamed columns
+  ## a big improvement but when I rationlize column names, see why these
+  ## empty column names aren't causing the columns to be dropped and fix it
+  out <- read_excel(test_sheet("skipping.xlsx"), skip = 10)
```
To make it less likely you'll forget about this column, you can use `skip()` to get a message in the tests

jennybc replied:
Done! Nice idea.
